src/main/scala/analysis/RegionInjector.scala

```diff
@@ -0,0 +1,317 @@
+                res ++= reducibleToRegion(binaryExpr, i)
+              case _ => // also treat as a region (for now) even if just Base + Offset without memLoad
+                res ++= exprToRegion(i.rhs, i)
+            }
+          }
+        }
+      case load: MemoryLoad => // treat as a region
+        res ++= exprToRegion(load.index, n)
+      case _ =>
+        evaluateExpressionWithSSA(expr, constantProp(n), n, reachingDefs).foreach { b =>
+          Logger.debug("BitVecLiteral: " + b)
+          val region = mmm.findDataObject(b.value)
+          if (region.isDefined) {
+            res += region.get
+          }
+        }
+    }
+    res
+  }
+
+  /** Default implementation of eval.
+   */
+  def eval(expr: Expr, cmd: Command): Expr = {
+    expr match
+      case literal: Literal => literal // ignore literals
+      case Extract(end, start, body) =>
+        Extract(end, start, eval(body, cmd))
+      case UninterpretedFunction(name, params, returnType) =>
+        val newParams = params.map { p => eval(p, cmd) }
+        UninterpretedFunction(name, newParams, returnType)
+      case Repeat(repeats, body) =>
+        Repeat(repeats, eval(body, cmd))
+      case ZeroExtend(extension, body) =>
+        ZeroExtend(extension, eval(body, cmd))
+      case SignExtend(extension, body) =>
+        SignExtend(extension, eval(body, cmd))
+      case UnaryExpr(op, arg) =>
+        UnaryExpr(op, eval(arg, cmd))
+      case BinaryExpr(op, arg1, arg2) =>
+        BinaryExpr(op, eval(arg1, cmd), eval(arg2, cmd))
+      case MemoryLoad(mem, index, endian, size) =>
+        // TODO: index should be replaced region
+        MemoryLoad(renameMemory(mem, index, cmd), eval(index, cmd), endian, size)
+      case variable: Variable => variable // ignore variables
+  }
+
+  def renameMemory(mem: Memory, expr: Expr, cmd : Command): Memory = {
+    val regions = exprToRegion(eval(expr, cmd), cmd)
+    if (regions.size == 1) {
+      Logger.warn(s"Mem CMD is: ${cmd}")
+      Logger.warn(s"Region found for mem: ${regions.head}")
+      regions.head match {
+        case stackRegion: StackRegion =>
+          return StackMemory(stackRegion.regionIdentifier, mem.addressSize, mem.valueSize)
+        case dataRegion: DataRegion =>
+          return SharedMemory(dataRegion.regionIdentifier, mem.addressSize, mem.valueSize)
+        case _ =>
+      }
+    } else if (regions.size > 1) {
+      Logger.warn(s"Mem CMD is: ${cmd}")
+      Logger.warn(s"Multiple regions found for mem: ${regions}")
+      mmm.mergeRegions(regions) match {
+        case stackRegion: StackRegion =>
+          return StackMemory(stackRegion.regionIdentifier, mem.addressSize, mem.valueSize)
+        case dataRegion: DataRegion =>
+          return SharedMemory(dataRegion.regionIdentifier, mem.addressSize, mem.valueSize)
+        case _ =>
+      }
+    } else {
+      Logger.warn(s"Mem CMD is: ${cmd}")
+      Logger.warn(s"No region found for mem")
+    }
+    mem
+  }
```

Comment:
Merging regions at the same time as we replace the `mem` variables in the IR will not work because it means that earlier variables that are replaced will not take into account any later merges.

For instance, if we have a program that the analysis determines accesses regions in the following order:
```
access region A
access region B
access region A or B
```
then when we encounter the access to region A or B, they should be merged into a single region, but the earlier accesses to A and B will not be updated to take into account that A and B are now the same region.

There needs to be two separate passes - a pass where regions are merged, and a pass where the memory variables are replaced.